### ISSUE TITLE
rename episodes

### ISSUE THREAD

**Issue by pavelbenisek** [brtmr/podfox#9]
Please add an option to rename episodes to real names (possibly shorten ones), then you could keep the filenames as links or symlinks.

I'm thinking of naming it as something like:
S01E03-real_name_of_the_episode.mp3

thanks

**Comment by tmajibon** [comment]
Just to clarify, are you wanting podfox to read ID3 tags to generate filenames?

It's possible, just requires additional libraries and some additional code centered around checking to see if the tag exists first.

**Comment by brtmr** [comment]
I assume he is referring to the title tag in the feed. I have not implemented a feature to do this, because podcasts are inconsistent about this.

**Comment by pavelbenisek** [comment]
It could be taken from the original RSS file too I guess. i'm not sure how
reliable the id3 tags would be - depending on source, but in RSS it must be
given always which should make it easier to implement; also all the
necessary libraries are already loaded.

**Comment by lv2eof** [comment]
Hi,
I made it by hand.
What did I do:
I downloaded all podcast files with "PodFox".
Then I moved all files but "feed.json" to another directory.
Then with the help of a spreadsheet, I renamed all files to the field "title" of the "json" file.
This way I can browse the directory where the files are and select the one I want to listen.
It didn't take too much time but it would be much more better if the program could do it for me.
But like we use to say: "who doesn't have dog hunts with the cat" :)
Cheers,
Luis

**Comment by pavelbenisek** [comment]
yep, of course the manual way is not the preferred one :)
I could write up something in bash/perl/php, but I am unable to manage it in python :(

Pavel

**Comment by brtmr** [comment]
Hmm... Let me explain why I did not go down this route - Mostly to avoid naming collisions. To convert every possible title (which might contain lots punctuation & unusual characters) to a filename, you have to reduce them to a sensible charset - which introduces the possibility that titles only differ in some small sense, like adding a question mark. These would be mapped to the same filename on disk. 

I therefore prefer to have the filename the original author chose, which may or may not be named in a consistent way.

I am gonna take some time later this week to properly implement a feature like this, which on a collision will either warn the user, or silently add a number to the end of the filename.

**Comment by tmajibon** [comment]
Simple collision avoidance option: generate a timestamp, either from the time in the feed, or a datetime.now().isoformat()?

**Comment by pavelbenisek** [comment]
You can simply add an episode number to the beginning of you human-readable-name, which has two advantages. 
1/ it distinguishes possible collisions
2/ it also helps when you listen (as I am) podcasts elsewhere using dumb player. - it gives you the right order.

thanks
Pavel

**Comment by pavelbenisek** [comment]
what about something like this for getting the unique names?

#!/usr/bin/python3

from os.path import expanduser
from sys import exit
import feedparser
import os
import os.path
import requests
import sys
from email.utils import parsedate
from dateutil.parser import parse
from datetime import datetime

url = 'http://www.bbc.co.uk/programmes/b00snr0w/episodes/downloads.rss'

feed = feedparser.parse(url)
for entry in feed['entries']:
    cdate = parse(entry['published'])
    print( cdate.strftime('%Y-%d-%m_%H:%M:%S') + "-" + os.path.basename(entry['link']))


thanks
Pavel

**Comment by tmajibon** [comment]
Yeah, that bit is basically what I was thinking. Just add some code to bypass it if there is no "published" tag. Also needs to be code for conflicting published tags too.

It's useful to recognize that the following is entirely possible:

- Two entries have the same published time
- Two entries have the same title
- There is no episode number attached

You also can't use it by position in the feed because many places use rolling feeds (older things are removed from the feed).

The simplest answer I can see is, the following in order:

1. Published date + converted title
2. Published date + file name (if the title's match, then they're clearly not informative)
3. Current Date + title
4. Current Date + file name
